src/devices_models/devices/common/add_to_expression.jl

```diff
@@ -1629,35 +1629,168 @@ function add_to_expression!(
-        for d in devices
-            name = PSY.get_name(d)
-            direction = 1.0
-            if specified_direction_mapping
-                if !haskey(direction_map, name)
-                    @warn "Direction not found for $(summary(d)). Will use the default from -> to direction"
-                else
-                    direction = direction_map[name]
-                end
+    contributing_devices_map = get_contributing_devices_map(model)
+    for map in NETWORK_REDUCTION_MAPS
+        network_reduction_map = all_branch_maps_by_type[map]
+        for branch_type in network_model.modeled_branch_types
+            if !haskey(contributing_devices_map, branch_type) ||
+               !haskey(network_reduction_map, branch_type)
+                continue
             end
-            for t in get_time_steps(container)
-                _add_to_jump_expression!(
-                    expression[service_name, t],
-                    variable[name, t],
-                    Float64(direction),
+            variable = get_variable(container, FlowActivePowerVariable(), branch_type)
+            contributing_devices = contributing_devices_map[branch_type]
+            for (arc_tuple, reduction_entry) in network_reduction_map[branch_type]
+                if _reduction_entry_in_interface(
+                    reduction_entry,
+                    contributing_devices,
                 )
+                    name = first(_get_branch_names(reduction_entry))
+                    if isempty(direction_map)
+                        direction = 1.0
+                    else
+                        direction = _get_direction(
+                            arc_tuple,
+                            reduction_entry,
+                            direction_map,
+                            network_reduction_data,
+                        )
+                    end
+                    for t in get_time_steps(container)
+                        _add_to_jump_expression!(
+                            expression[service_name, t],
+                            variable[name, t],
+                            Float64(direction),
+                        )
+                    end
+                end
             end
         end
     end
     return
 end
 
+function _get_direction(
+    ::Tuple{Int, Int},
+    reduction_entry::PSY.ACTransmission,
+    direction_map::Dict{String, Int},
+    ::PNM.NetworkReductionData,
+)
+    name = PSY.get_name(reduction_entry)
+    if !haskey(direction_map, name)
+        @warn "Direction not found for $(summary(d)). Will use the default from -> to direction"
+        return 1.0
+    else
+        return direction_map[name]
+    end
+end
+
+function _get_direction(
+    arc_tuple::Tuple{Int, Int},
+    reduction_entry::Set{PSY.ACTransmission},
+    direction_map::Dict{String, Int},
+    network_reduction_data::PNM.NetworkReductionData,
+)
+    directions = [
```

Comment:
This code can be simplified into a single loop instead of two. If there are only a few items in each set it should be fine but leave a comment that this is fine because there aren't like 10_000 parallel branches.